### DISCUSSION TITLE
History substring search

### DISCUSSION THREAD

**Discussion by cornerot** [nushell/nushell#7968] [Ideas]
Fish shell history search feature, where you can type in any part (not only by prefix) of any command from history and then press chosen keys, such as the UP and DOWN arrows, to cycle through matches.

**Comment by fdncred** [comment]
We do have `$env.config.completion.algorithm` that has `prefix` and `fuzzy` as options. Does this do what you're looking for?

**Reply by cornerot** [reply]
No. This option only affects commands autocompletion, not a history.
For example [fish](https://asciinema.org/a/HLiIPEVXQFbtQ2w8o1bjyw9N3) and [nu](https://asciinema.org/a/M0vP3Ka90ZXShNFHz5fsDJXBx) behavior

**Reply by fdncred** [reply]
Hmmm, apparently we don't support this yet.

**Reply by cornerot** [reply]
It seems reedline issue

**Comment by sommerper** [comment]
Is there a way to do this with https://www.nushell.sh/book/line_editor.html#keybindings
Basically detecting arrow up / down and searching the history?

**Reply by fdncred** [reply]
today, if i type `str<up-arrow>` I get a search based on the prefix. So, I'm not really sure what the question is. it seems to work.

**Reply by sommerper** [reply]
It's true that the prefix search works - the issue is that - let's say I have previously written:
```cat news.txt```

... I can write `cat` in the prompt and use arrow up to scroll through all entries that have `cat` as the first part of the string.

But if I want to see all the history entries where I do something with `news.txt`, write that into the prompt, and press arrow up it will find nothing.

So it wont find something like:
```
touch news.txt
nano news.txt
cat news.txt
```

It's something that I do quiet often - but maybe I'm alone in doing that... =)

**Reply by fdncred** [reply]
With the default_config.nu, it seems like ctrl+r does this. There's also reverse serach with ctrl+q. I also have several history_menus in my config.nu.

**Reply by sommerper** [reply]
Yes, ctrl-q does exactly this.

I think what OP and I would like is that this is triggered on arrow up and down (with the current text in the prompt), without having to press ctrl-q first and then writing the text afterwards.

I often find myself having already typed the filename that I want to work with and then browse through the history.

![rec](https://github.com/nushell/nushell/assets/1512048/1164024e-3771-4a0c-98e1-e19bb607e13b)

It might seem a bit unnecessary but it is a slightly faster procedure if you know that what you're searching for is not that far back in history.

So this isn't a critique or anything but rather an optimization of sorts.

**Reply by fdncred** [reply]
ok, search in your config.nu and look for ctrl+q and change that to up. play around with the keybindings to see if you can get it working the way you want. `keybindings list` should show you what is available.

**Reply by sommerper** [reply]
I looked into it but I don't think I can do it without being able to execute custom commands.

If the prompt is empty it should just go one step up and down in history - but if something is already typed then it should switch to do a history substring search.

**Comment by MilesCranmer** [comment]
https://github.com/nushell/nushell/pull/12616

@cornerot @fdncred @sommerper

**Comment by ibrokemypie** [comment]
Any updates on this? Really missing https://github.com/zsh-users/zsh-history-substring-search, especially if I want to get back to a previous ssh command, but the ssh is prefixed with AWS_PROFILE envvars or something like that, current up bind wont find it forcing me to go into the control+r search instead which isn't as ergonomic.

**Reply by fdncred** [reply]
https://github.com/nushell/nushell/pull/15511

**Reply by roastedramen** [reply]
I set my completion algorithm to substring as indicated by that change but it hasn't seemed to have changed anything, am I missing something?
<img width="916" height="348" alt="image" src="https://github.com/user-attachments/assets/9e2ca843-3212-4bec-9d81-59ef5bae7732" />

**Reply by fdncred** [reply]
@ysthakur is there something special you have to do to get substring matching to work?

**Reply by ysthakur** [reply]
I'm afraid that for history, we use a completer baked into Reedline, so the Nushell config doesn't currently do anything. 

However, I do see different search types here: https://github.com/nushell/reedline/blob/main/src%2Fhistory%2Fbase.rs#L7-L29. I'll check where that's set, so that Nushell can use env.config.completions.algorithm to set the right search type in Reedline's history completer.

**Reply by ysthakur** [reply]
Seems like we hardcode history navigation modes in Reedline: https://github.com/nushell/reedline/blob/405299b94b4cde1665d509f9e2ae9f5048fa5a76/src/engine.rs#L1357-L1371

This should be configurable in some way in both Reedline and Nushell, but it'll take some thinking about design.

**Comment by Ocyss** [comment]
Hi @ysthakur @fdncred, just bumping this thread to see if there have been any new developments or thoughts on the Reedline configuration design you mentioned last year? 

I recently migrated from zsh and absolutely love Nushell, but I really miss `zsh-history-substring-search`. It's a massive quality-of-life feature for handling long, prefixed commands. Would love to know if this is still on the radar or if there's any way the community can help move it forward!

**Reply by fdncred** [reply]
We're probably accept a reedline and nushell PR to help with this.
